Change error message for type param which is a non-isbits value

dagbr opened the pull request:
Fixes #39786 by using Type{:TypeParameter} as the expected type in a TypeError
to indicate a non-valid type parameter.

dbrueg committed:
Change error message for type param which is a non-isbits Number

Fixes #39786 by using Type{:TypeParameter} as the expected type in a TypeError
to indicate an invalid type parameter.

JeffBezanson commented:
I think this could use some discussion. I'm not sure it should only apply to Numbers. And the implementation is clever but I think a bit too obscure.

dbrueg committed:
Change error message for type param which is a non-isbits value

Fixes #39786

dbrueg committed:
Revert "Change error message for type param which is a non-isbits Number"

This reverts commit badd1599625201e6f7e2a432f040423a22ae9383.

dagbr commented:
@JeffBezanson Here is another implementation, which does not use a symbol inside `Type{ }`. It detects the origin of the exception using `ex.func` and `ex.context` instead. It is not specific to Numbers anymore.

JeffBezanson commented on base/errorshow.jl:
```diff
@@ -64,12 +64,16 @@ function showerror(io::IO, ex::TypeError)
             targs = ("Type{", ex.got, "}")
         else
             targs = ("a value of type $(typeof(ex.got))",)
+            if ex.func == Symbol("Type") && ex.context == "parameter"
```
```suggestion
            if ex.func === :Type && ex.context == "parameter"
```

JeffBezanson commented:
How about this format:
```
TypeError: values of type BigInt are not valid as type parameters. Valid type parameters are
    - Types
    - Bits values
    - Symbols
    - Tuples of symbols or bits values
 ```

dbrueg committed:
Update base/errorshow.jl

Co-authored-by: Jeff Bezanson <jeff.bezanson@gmail.com>

dbrueg committed:
More detailed error message for values which are invalid as type parameters.

dagbr commented:
@JeffBezanson I have changed the error message to match your suggestion. Should I squash?

simeonschaub commented on base/errorshow.jl:
```diff
@@ -78,7 +82,17 @@ function showerror(io::IO, ex::TypeError)
+                          - Types
+                          - Bits values
+                          - Symbols
+                          - Tuples of symbols or bits values
```
```suggestion
                          - Tuples of symbols, bits values or other tuples thereof
```

vtjnash replied:
It is not recursive:
```
julia> Tuple{((:a,),)}
ERROR: TypeError: in Tuple, in parameter, expected Type, got a value of type Tuple{Tuple{Symbol}}
Stacktrace:
 [1] top-level scope
   @ REPL[2]:1
```

simeonschaub replied:
Interesting, it does work with ints though:
```julia
julia> Tuple{((1,),)}
Tuple{((1,),)}
```

simeonschaub replied:
```suggestion
                          - Tuples of symbols, bits values or other tuples of bits values
```

vtjnash replied:
That is an isbits value

simeonschaub replied:
Ah, I see. Since we explicitly list tuples here, perhaps
```suggestion
                          - Tuples of symbols or bits values, including tuples of bits values
```
would clarify this a bit?

dagbr replied:
how about just "- tuples of symbols"? (see the new commit)

simeonschaub reviewed (approved):
One tiny correction, otherwise LGTM

dbrueg committed:
simplify error message

remove tuple of bits values since they are bits values

lowercase because "-Tuples" might be interpreted to refer to `Tuple{...}`

vtjnash commented on base/errorshow.jl:
```diff
@@ -78,7 +82,17 @@ function showerror(io::IO, ex::TypeError)
+            print(io, """
+                      values of type $(typeof(ex.got)) are not valid as type parameters. Valid type parameters are
+                          - types
+                          - bits values
+                          - symbols
+                          - tuples of symbols
```
```suggestion
                          - isbits values
                          - symbols
                          - tuples of symbols, types, or isbits values
```
